src/ray/common/file_system_monitor.cc

```diff
@@ -0,0 +1,130 @@
+      io_context_(),
+      monitor_thread_([this] {
+        /// The asio work to keep io_contex_ alive.
+        boost::asio::io_service::work io_service_work_(io_context_);
+        io_context_.run();
+      }),
+      runner_(io_context_) {
+  runner_.RunFnPeriodically([this] { over_capacity_ = CheckIfAnyPathOverCapacity(); },
+                            monitor_interval_ms,
+                            "FileSystemMonitor.CheckIfAnyPathOverCapacity");
+}
+
+FileSystemMonitor::~FileSystemMonitor() {
+  io_context_.stop();
+  if (monitor_thread_.joinable()) {
+    monitor_thread_.join();
+  }
+}
+
+std::optional<std::filesystem::space_info> FileSystemMonitor::Space(
+    const std::string &path) const {
+  std::error_code ec;
+  const std::filesystem::space_info si = std::filesystem::space(path, ec);
+  if (ec) {
+    RAY_LOG_EVERY_MS(WARNING, 60 * 1000)
+        << "Failed to get capacity of " << path << " with error: " << ec.message();
+    return std::nullopt;
+  }
+  return si;
+}
+
+bool FileSystemMonitor::OverCapacity() const { return over_capacity_.load(); }
+
+bool FileSystemMonitor::CheckIfAnyPathOverCapacity() const {
+  if (paths_.empty()) {
+    return false;
+  }
+
+  if (capacity_threshold_ == 0) {
+    return true;
+  }
+
+  if (capacity_threshold_ >= 1) {
+    return false;
+  }
+
+  for (auto &path : paths_) {
+    if (OverCapacityImpl(path, Space(path))) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool FileSystemMonitor::OverCapacityImpl(
+    const std::string &path,
+    const std::optional<std::filesystem::space_info> &space_info) const {
+  if (!space_info.has_value()) {
+    return false;
+  }
+  if (space_info->capacity <= 0) {
+    RAY_LOG_EVERY_MS(ERROR, 60 * 1000) << path << " has no capacity.";
+    return true;
+  }
+
+  if ((1 - 1.0f * space_info->available / space_info->capacity) < capacity_threshold_) {
+    return false;
+  }
+
+  RAY_LOG_EVERY_MS(ERROR, 10 * 1000)
+      << path << " is over capacity, available: " << space_info->available
+      << ", capacity: " << space_info->capacity << ", threshold: " << capacity_threshold_;
+  return true;
+}
+
+std::vector<std::string> ParseSpillingPaths(const std::string &spilling_config) {
+  std::vector<std::string> spilling_paths;
+
+  try {
+    json spill_config = json::parse(spilling_config);
+    const auto &directory_path = spill_config.at("params").at("directory_path");
+    if (directory_path.is_string()) {
+      spilling_paths.push_back(directory_path);
+    } else if (directory_path.is_array()) {
+      for (auto &entry : directory_path) {
+        if (entry.is_string()) {
+          spilling_paths.push_back(entry);
+        } else {
+          RAY_LOG(ERROR) << "Failed to parse spilling path: " << entry;
+        }
+      }
+    } else {
+      RAY_LOG(ERROR) << "Failed to parse spilling path: " << directory_path;
+    }
+  } catch (json::exception &ex) {
+    RAY_LOG(ERROR) << "Failed to load spilling config: " << ex.what()
```

Comment:
Since these errors are user-facing, it would be good to add a short explanation of what could go wrong.